File: Rosu/programmazione/lezioni/java/dynamic-programming/src/ro/razvan/uniud/memoization/java/LongestCommonSubsequence.java
package ro.razvan.uniud.memoization.java;

@SuppressWarnings("Duplicates")
class LongestCommonSubsequence {

    private static final int UNKNOWN = -1;

    static int llcsRec(final String u, final String v) {
        if (u.isEmpty() || v.isEmpty()) {
            return 0;
        } else if (u.charAt(0) == v.charAt(0)) {
            return 1 + llcsRec(u.substring(1), v.substring(1));
        } else {
            return Math.max(
                    llcsRec(u.substring(1), v),
                    llcsRec(u, v.substring(1))
            );
        }
    }

    static int llcsMemRec(final String u, final String v) {
        final var matrix = new int[u.length() + 1][v.length() + 1];

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                matrix[i][j] = UNKNOWN;
            }
        }

        return llcsMemRec(u, v, matrix);
    }

    private static int llcsMemRec(final String u, final String v, final int[][] matrix) {

        final var i = u.length();
        final var j = v.length();

        if (matrix[i][j] == UNKNOWN) {

            if (u.isEmpty() || v.isEmpty()) {
                matrix[i][j] = 0;
            } else if (u.charAt(0) == v.charAt(0)) {
                matrix[i][j] = llcsMemRec(u.substring(1), v.substring(1)) + 1;
            } else {
                matrix[i][j] = Math.max(
                        llcsMemRec(u.substring(1), v, matrix),
                        llcsMemRec(u, v.substring(1), matrix)
                );
            }

        }

        return matrix[i][j];
    }

    static int llcsIter(final String u, final String v) {
        final var m = u.length();
        final var n = v.length();
        final var matrix = new int[m + 1][n + 1];

        for (int i = 0; i < matrix.length; i++) {
            matrix[i][0] = 0;
        }

        for (int i = 1; i <= n; i++) {
            matrix[0][i] = 0;
        }

        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[i].length; j++) {
                if (u.charAt(m - i) == v.charAt(n - j)) {
                    matrix[i][j] = 1 + matrix[i - 1][j - 1];
                } else {
                    matrix[i][j] = Math.max(matrix[i][j - 1], matrix[i - 1][j]);
                }
            }
        }

        return matrix[m][n];
    }


    static String lcs(final String u, final String v) {
        final var m = u.length();
        final var n = v.length();
        final var matrix = new int[m + 1][n + 1];

        for (int i = 0; i < matrix.length; i++) {
            matrix[i][0] = 0;
        }

        for (int i = 1; i <= n; i++) {
            matrix[0][i] = 0;
        }

        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[i].length; j++) {
                if (u.charAt(m - i) == v.charAt(n - j)) {
                    matrix[i][j] = 1 + matrix[i - 1][j - 1];
                } else {
                    matrix[i][j] = Math.max(
                            matrix[i - 1][j],
                            matrix[i][j - 1]
                    );
                }
            }
        }

        final var builder = new StringBuilder();
        var i = m;
        var j = n;

        while (matrix[i][j] != 0) {
            if (u.charAt(m - i) == v.charAt(n - j)) {
                builder.append(u.charAt(m - i));
                i--;
                j--;
            } else if (matrix[i - 1][j] < matrix[i][j - 1]) {
                j--;
            } else {
                i--;
            }
        }

        return builder.toString();
    }

}

File: Rosu/programmazione/lezioni/java/dynamic-programming/src/ro/razvan/uniud/memoization/kotlin/LongestCommonSubsequence.kt
package ro.razvan.uniud.memoization.kotlin

import kotlin.math.max

object LongestCommonSubsequence {

    private const val UNKNOWN = -1

    fun llcsRec(u: String, v: String): Int = when {
        u.isEmpty() || v.isEmpty() -> 0
        u[0] == v[0] -> 1 + llcsRec(u.substring(1), v.substring(1))
        else -> max(
            llcsRec(u.substring(1), v),
            llcsRec(u, v.substring(1))
        )
    }

    fun llcsMemRec(
        u: String,
        v: String
    ): Int = llcsMemRec(u, v, Array(u.length + 1) { IntArray(v.length + 1) { UNKNOWN } })

    private fun llcsMemRec(
        u: String,
        v: String,
        matrix: Array<IntArray>
    ): Int {

        if (matrix[u.length][v.length] == UNKNOWN) {

            matrix[u.length][v.length] = when {
                u.isEmpty() || v.isEmpty() -> 0
                u[0] == v[0] -> 1 + llcsMemRec(u.substring(1), v.substring(1), matrix)
                else -> max(
                    llcsMemRec(u.substring(1), v),
                    llcsMemRec(u, v.substring(1))
                )
            }

        }

        return matrix[u.length][v.length]
    }

    fun llcsIter(u: String, v: String): Int {

        val m = u.length
        val n = v.length
        val matrix = Array(m + 1) { i ->
            IntArray(n + 1) { j ->
                if (j == 0 || i == 0) {
                    0
                } else {
                    UNKNOWN
                }
            }
        }

        for (i in 1 until matrix.size) {

            for (j in 1 until matrix[i].size) {

                matrix[i][j] = when {

                    u.isEmpty() || v.isEmpty() -> 0

                    u[m - i] == v[n - j] -> 1 + matrix[i - 1][j - 1]

                    else -> max(
                        matrix[i - 1][j],
                        matrix[i][j - 1]
                    )

                }

            }

        }

        return matrix[m][n]
    }

    fun lcs(u: String, v: String): String {
        val m = u.length
        val n = v.length
        val matrix = Array(m + 1) { i ->
            IntArray(n + 1) { j ->
                if (j == 0 || i == 0) {
                    0
                } else {
                    UNKNOWN
                }
            }
        }

        for (i in 1 until matrix.size) {

            for (j in 1 until matrix[i].size) {
                matrix[i][j] = when {

                    u.isEmpty() || v.isEmpty() -> 0

                    u[m - i] == v[n - j] -> 1 + matrix[i - 1][j - 1]

                    else -> max(
                        matrix[i - 1][j],
                        matrix[i][j - 1]
                    )

                }
            }

        }

        // fine llcs

        val builder = StringBuilder()
        var i = m
        var j = n

        while (matrix[i][j] != 0) {

            when {

                u[m - i] == v[n - j] -> {
                    builder.append(u[m - i])
                    j--
                    i--
                }

                matrix[i - 1][j] < matrix[i][j - 1] -> j--
                else -> i--

            }

        }

        return builder.toString()
    }


}